### PR TITLE
Feature/bngp 188 legal agreement preupload

### PR DESCRIPTION
**Pull request by reneli5**

### CODE REVIEW

**Comment by teddmason** [comment]
@reneli5 This branch needs rebasing with master. A lot of the code changes reported in the pull request are changes already committed to master. Unsure how this has come about, regarding merging strategy

File: .gitignore [hunk 1]
```diff
@@ -121,5 +121,6 @@ packages/**/babel.config.json
 .jest-with-debug/
 #intelij management files
 .idea
+.scannerwork
```
Comment (pwadmore-ea) [inline .gitignore:124]: I'd be interested to know how your local code quality checks have been set up. Things like this should benefit the whole team. Maybe you could do a show and tell at one of the tech team calls or direct others in the team to resources you used?

File: .scannerwork/report-task.txt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+projectKey=DEFRA_biodiversity-net-gain-service
+serverUrl=http://127.0.0.1:9000
+serverVersion=9.6.1.59531
+dashboardUrl=http://127.0.0.1:9000/dashboard?id=DEFRA_biodiversity-net-gain-service
+ceTaskId=AYPNI6HjXAHgg2375MjL
+ceTaskUrl=http://127.0.0.1:9000/api/ce/task?id=AYPNI6HjXAHgg2375MjL
```
Comment (pwadmore-ea) [inline .scannerwork/report-task.txt:6]: Does this file and/or .scannerwork/.sonarlock contain any sensitive values that require the file to be removed from version control history and the regeneration of sensitive values accordingly?

File: packages/application-to-register-webapp/client/js/pages/land/add-legal-agreement-parties.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+      $('.govuk-radios__item').click((event) => {
+        if ($(event.target.parentNode).next('div').hasClass('govuk-radios__conditional--hidden')) {
+          $(event.target.parentNode).next('div').removeClass('govuk-radios__conditional--hidden')
+        } else {
+          $(event.target.parentNode).parent().show().find('.govuk-radios__conditional').addClass('govuk-radios__conditional--hidden')
+        }
+      })
+    })
+  }
+})
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/client/js/pages/land/add-legal-agreement-parties.js:14]: Is this file supposed to be referenced from [packages/application-to-register-webapp/src/views/land/add-legal-agreement-parties.html](https://github.com/DEFRA/biodiversity-net-gain-service/blob/a471b0c0806c3658c3e62e598636980effbf6439/packages/application-to-register-webapp/src/views/land/add-legal-agreement-parties.html)?.

If so please make corresponding changes to the HTML file and [packages/application-to-register-webapp/webpack.config.js](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/webpack.config.js)

File: packages/application-to-register-webapp/src/routes/__tests__/land/add-legal-agreement-parties.spec.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    })
+
+    it('should fail to add single legal party to legal agreement without organisation name', async () => {
+      postOptions.payload = {
+        'organisation[0][organisationName]': '',
+        'organisation[0][role]': 'Developer',
+        otherPartyName: ''
+      }
+      const response = await submitPostRequest(postOptions, 200)
+      expect(response.statusCode).toBe(200)
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/add-legal-agreement-parties.spec.js:97]: Please can you add additional checks for expected errors. [For example](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/src/routes/__tests__/land/add-landowners.spec.js)

File: packages/application-to-register-webapp/src/routes/__tests__/land/add-legal-agreement-parties.spec.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/add-legal-agreement-parties.spec.js:107]: Please can you add additional checks for expected errors. [For example](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/src/routes/__tests__/land/add-landowners.spec.js).

File: packages/application-to-register-webapp/src/routes/__tests__/land/add-legal-agreement-parties.spec.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    })
+
+    it('should fail to add single legal party to legal agreement without organisation name nor role', async () => {
+      postOptions.payload = {
+        'organisation[0][organisationName]': '',
+        otherPartyName: ''
+      }
+      const response = await submitPostRequest(postOptions, 200)
+      expect(response.statusCode).toBe(200)
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/add-legal-agreement-parties.spec.js:116]: Please can you add additional checks for expected errors. [For example](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/src/routes/__tests__/land/add-landowners.spec.js).

File: packages/application-to-register-webapp/src/routes/__tests__/land/add-legal-agreement-parties.spec.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        'organisation[0][role]': 'Other',
+        otherPartyName: [
+          '',
+          ''
+        ],
+        'organisation[1][organisationName]': 'Test Two',
+        'organisation[1][role]': 'Other'
+      }
+      const response = await submitPostRequest(postOptions, 200)
+      expect(response.statusCode).toBe(200)
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/add-legal-agreement-parties.spec.js:131]: Please can you add additional checks for expected errors. [For example](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/src/routes/__tests__/land/add-landowners.spec.js).

File: packages/application-to-register-webapp/src/routes/__tests__/land/check-legal-agreement-details.spec.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
+import constants from '../../../utils/constants.js'
+
+const url = '/land/check-legal-agreement-details'
+const mockDataPath = 'packages/application-to-register-webapp/src/__mock-data__/uploads/legal-agreements'
+
+describe(url, () => {
+  const rdisMap = new Map()
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/check-legal-agreement-details.spec.js:7]: redisMap typo.

File: packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+      postOptions = {
+        url,
+        payload: {}
+      }
+    })
+
+    it('should add a legal agreement start date', async () => {
+      postOptions.payload['legalAgreementStartDate-day'] = '01'
+      postOptions.payload['legalAgreementStartDate-month'] = '01'
+      postOptions.payload['legalAgreementStartDate-year'] = '1971'
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js:26]: Although, not part of the acceptance criteria I question the service accepting start dates for legal agreements (and other documents) before dates specified by legislation and/or policy. I think this needs confirmation with the product owner.

File: packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+      expect(response.statusCode).toBe(200)
+    })
+
+    it('should fail to add a legal agreement start date with empty day', async () => {
+      postOptions.payload['legalAgreementStartDate-day'] = ''
+      postOptions.payload['legalAgreementStartDate-month'] = '01'
+      postOptions.payload['legalAgreementStartDate-year'] = '2022'
+
+      const response = await submitPostRequest(postOptions, 200)
+      expect(response.statusCode).toBe(200)
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js:69]: Please can you add additional checks for expected errors. [For example](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/src/routes/__tests__/land/add-landowners.spec.js).

File: packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    })
+
+    it('should fail to add a legal agreement start date with empty month', async () => {
+      postOptions.payload['legalAgreementStartDate-day'] = '01'
+      postOptions.payload['legalAgreementStartDate-month'] = ''
+      postOptions.payload['legalAgreementStartDate-year'] = '2022'
+
+      const response = await submitPostRequest(postOptions, 200)
+      expect(response.statusCode).toBe(200)
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js:78]: Please can you add additional checks for expected errors. [For example](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/src/routes/__tests__/land/add-landowners.spec.js).

File: packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    })
+
+    it('should fail to add a legal agreement start date with empty year', async () => {
+      postOptions.payload['legalAgreementStartDate-day'] = '01'
+      postOptions.payload['legalAgreementStartDate-month'] = '01'
+      postOptions.payload['legalAgreementStartDate-year'] = ''
+
+      const response = await submitPostRequest(postOptions, 200)
+      expect(response.statusCode).toBe(200)
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js:87]: Please can you add additional checks for expected errors. [For example](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/src/routes/__tests__/land/add-landowners.spec.js).

File: packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+    })
+
+    it('should fail to add a legal agreement start date with invalid dates', async () => {
+      postOptions.payload['legalAgreementStartDate-day'] = 'xx'
+      postOptions.payload['legalAgreementStartDate-month'] = 'zz'
+      postOptions.payload['legalAgreementStartDate-year'] = 'cc'
+
+      const response = await submitPostRequest(postOptions, 200)
+      expect(response.statusCode).toBe(200)
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js:96]: Please can you add additional checks for expected errors. [For example](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/src/routes/__tests__/land/add-landowners.spec.js).

File: packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    })
+
+    it('should fail to add a legal agreement start date with wrong month', async () => {
+      postOptions.payload['legalAgreementStartDate-day'] = '11'
+      postOptions.payload['legalAgreementStartDate-month'] = '23'
+      postOptions.payload['legalAgreementStartDate-year'] = '2020'
+
+      const response = await submitPostRequest(postOptions, 200)
+      expect(response.statusCode).toBe(200)
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js:114]: Please can you add additional checks for expected errors. [For example](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/src/routes/__tests__/land/add-landowners.spec.js).

File: packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    })
+
+    it('should fail to add a legal agreement start date with wrong year', async () => {
+      postOptions.payload['legalAgreementStartDate-day'] = '11'
+      postOptions.payload['legalAgreementStartDate-month'] = '23'
+      postOptions.payload['legalAgreementStartDate-year'] = '2a20'
+
+      const response = await submitPostRequest(postOptions, 200)
+      expect(response.statusCode).toBe(200)
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js:123]: Please can you add additional checks for expected errors. [For example](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/src/routes/__tests__/land/add-landowners.spec.js).

File: packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-type.spec.js [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+        const h = {
+          view: (view, context) => {
+            viewResult = view
+            contextResult = context
+          }
+        }
+        await legalAgreementDetails.default[0].handler(request, h)
+        expect(viewResult).toEqual('land/legal-agreement-type')
+        expect(contextResult.dontHave).toEqual(true)
+      })
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-type.spec.js:80]: Is this test valid?

If somebody has indicated that they do not have a legal agreement they are redirected to a page that says they need one. As such I believe you should not be able to reach the legal agreement page using a cached value of 'I do not have a legal agreement'.

File: packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js [hunk 17]
```diff
@@ -0,0 +1,177 @@
+import constants from '../../utils/constants.js'
+
+const START_ID = '#organisation'
+let referredFrom
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js:4]: I do not think module level variables are safe for concurrent processing. While this approach will work for single client usage such as local development and unit testing, concurrent processing runs the risk of a session using referrer values from another session.

@teddmason  is working on an alternative that stores referrer data in the Redis cache so that it is scoped to a particular session. I suggest liaising with @teddmason to introduce the functionality to this pull request. Associated unit tests might need refactoring accordingly.

File: packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+function processUndefinedRole (partySelectionData, index, combinedError) {
+  partySelectionData.roleError.push({
+    text: 'Select the role',
+    href: `[${START_ID}[${index}][role]`,
+    index
+  })
+  combinedError.push({
+    text: 'Select the role',
+    href: `[${START_ID}[${index}][role]`
+  })
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js:43]: I suggest extracting common JSON into a variable rather than copy and paste.

File: packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+      otherParty = ''
+      partySelectionData.roleError.push({
+        text: 'Other type of role cannot be left blank',
+        href: `[${START_ID}[${index}][role]`,
+        index
+      })
+      combinedError.push({
+        text: 'Other type of role cannot be left blank',
+        href: `[${START_ID}[${index}][role]`
+      })
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js:61]: I suggest extracting common JSON into a variable rather than copy and paste.

File: packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+      break
+    case 'Other':
+      roleDetails = {
+        organisationIndex: indexValue,
+        rowIndex: 4,
+        other: true
+      }
+      break
+    default:
+      roleDetails = undefined
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js:135]: Missing break?

File: packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+function processEmptyPartySelection (partySelectionData, index, combinedError) {
+  partySelectionData.organisationError.push({
+    text: 'Enter the name of the legal party',
+    href: `[${START_ID}[${index}][organisationName]`,
+    index
+  })
+  combinedError.push({
+    text: 'Enter the name of the legal party',
+    href: `[${START_ID}[${index}][organisationName]`
+  })
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js:14]: I suggest extracting common JSON into a variable rather than copy and paste.

File: packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js (hunk 18 above)
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js:43]: I suggest extracting common JSON into a variable rather than copy and paste.

File: packages/application-to-register-webapp/src/routes/land/legal-agreement-start-date.js [hunk 22]
```diff
@@ -0,0 +1,47 @@
+import constants from '../../utils/constants.js'
+import moment from 'moment'
+import { dateClasses, validateDate } from '../../utils/helpers.js'
+
+const ID = 'legalAgreementStartDate'
+let referredFrom
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/land/legal-agreement-start-date.js:6]: I do not think module level variables are safe for concurrent processing. While this approach will work for single client usage such as local development and unit testing, concurrent processing runs the risk of a session using referrer values from another session.

@teddmason  is working on an alternative that stores referrer data in the Redis cache so that it is scoped to a particular session. I suggest liaising with @teddmason to introduce the functionality to this pull request. Associated unit tests might need refactoring accordingly.

File: packages/application-to-register-webapp/src/routes/land/upload-legal-agreement.js [hunk 23]
```diff
@@ -5,7 +5,7 @@ import constants from '../../utils/constants.js'
 import { uploadFiles } from '../../utils/upload.js'
 
 const LEGAL_AGREEMENT_ID = '#legalAgreement'
-
+let referredFrom
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/land/upload-legal-agreement.js:8]: I do not think module level variables are safe for concurrent processing. While this approach will work for single client usage such as local development and unit testing, concurrent processing runs the risk of a session using referrer values from another session.

@teddmason  is working on an alternative that stores referrer data in the Redis cache so that it is scoped to a particular session. I suggest liaising with @teddmason to introduce the functionality to this pull request. Associated unit tests might need refactoring accordingly.

File: packages/application-to-register-webapp/src/views/land/add-legal-agreement-parties.html [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+                                {% endfor%}
+                                {% call govukFieldset({
+                                    classes: 'moj-add-another__item',
+                                    legend: {
+                                        text: 'Legal party',
+                                        classes: 'moj-add-another__title govuk-fieldset__legend--m',
+                                        isPageHeading: false
+                                    }
+                                }) %}
+                    <form method="post">
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/views/land/add-legal-agreement-parties.html:54]: Incorrect indentation

File: packages/application-to-register-webapp/src/views/land/add-legal-agreement-parties.html [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+                                    {% endif %}
+                                {% endcall %}
+                            {% endfor %}
+{#                        {% endif %}#}
+                        <div class="moj-button-action">
+                            {{ govukButton({
+                                text: 'Add another',
+                                classes: 'govuk-button--secondary moj-add-another__add-button govuk-!-margin-bottom-4'
+                            }) }}
+                        </div>
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/views/land/add-legal-agreement-parties.html:174]: Check formatting/indentation

File: packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-type.spec.js [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+        }
+        const h = {
+          view: (view, context) => {
+            viewResult = view
+            contextResult = context
+          }
+        }
+        await legalAgreementDetails.default[0].handler(request, h)
+        expect(viewResult).toEqual('land/legal-agreement-type')
+        expect(contextResult.dontHave).toEqual(true)
```
Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-type.spec.js:79]: Is this test temporary until the kickout functionality reflects the prototype?

If somebody has indicated that they do not have a legal agreement the prototype redirects to a page that says they need one. As such I believe you should not be able to reach the legal agreement page using a cached value of 'I do not have a legal agreement'.

Comment (pwadmore-ea) [inline packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-type.spec.js:79]: Is this test temporary until the kickout functionality reflects the prototype?

If somebody has indicated that they do not have a legal agreement the prototype redirects to a page that says they need one. As such I believe you should not be able to reach the legal agreement page using a cached value of 'I do not have a legal agreement'.

File: .gitignore (hunk 1 above)
Comment (reneli5) [reply .gitignore:124]: .idea is my IDE. I normally remove it from commit, but this time I failed to notice it. I have removed it in the subsequent commit by adding it to .gitignore.
.scannerwork helps me run sonar scanner before I commit. This way, I remove sonar noise from commit. It does not work 100%

File: .scannerwork/report-task.txt (hunk 2 above)
Comment (reneli5) [reply .scannerwork/report-task.txt:6]: No.

**Review by pwadmore-ea — CHANGES_REQUESTED** [review]
@reneli5 - I have made a number of comments on individual lines and blocks of lines. Some of them are repeated where the same comment applies in multiple files.

Additionally:

- I recommend that the add legal agreement parties view is supplemented with page level JavaScript to enhance some of the existing MOJ 'Add Another'  component functionality (for example, ensuring that error messages are not propagated to a new item when 'Add Another' is clicked). @teddmason has created some [page level scripting](https://github.com/DEFRA/biodiversity-net-gain-service/blob/master/packages/application-to-register-webapp/client/js/pages/land/add-landowners.js) for adding landowners to enhance the default functionality that could be adapted for similar pages. Rather than copying, pasting and adapting the JavaScript that has been created. I'm wondering if the JavaScript can be made more generic (such as through parameterisation) so that any enhancements/bug fixes will be picked by all pages requiring 'Add Another' functionality automatically..  
- Bug - If I Add Another legal agreement party the visibility of the text field for entering a custom party name appears to break such that component visibility does not respond to changes in the party type until the page is reloaded. I can demonstrate if needed.
- Please correct the typo legalAgrement throughout.

Apologies for the number of comments. Please let me know if anything is unclear.

File: packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js (hunk 17 above)
Comment (reneli5) [reply packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js:4]: I agree. I will liaise with Ted on this.

File: packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-type.spec.js (hunk 26 above)
Comment (reneli5) [reply packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-type.spec.js]: This is temporary until the prototype is reflected when the other page is implemented.

File: packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js (hunk 9 above)
Comment (reneli5) [reply packages/application-to-register-webapp/src/routes/__tests__/land/legal-agreement-start-date-spec.js:26]: I will bring this up with Shawn

File: packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js (hunk 20 above)
Comment (reneli5) [reply packages/application-to-register-webapp/src/routes/land/add-legal-agreement-parties.js:150]: No need to break on default

**Review by pwadmore-ea — CHANGES_REQUESTED** [review]
This bug is still present - If I Add Another legal agreement party the visibility of the text field for entering a custom party name appears to break such that component visibility does not respond to changes in the party type until the page is reloaded.

Clarification - This bug seems to be intermittent.

Additionally, when re-entering the add legal agreement parties page from the check your answers page, the stored role is replaced with undefined.

**Review by pwadmore-ea — APPROVED** [review]
Unit tests requiring SignalR negotiation are failing due to Azure Functions Core Tools failing to start currently. This appears to be due to a [broken URL associated with extension bundles](https://github.com/Azure/azure-functions-core-tools/issues/3179).

Emulation is used in a limited number of unit tests due to difficulty in mocking all required elements.

A administrator forced merge to master is proposed to unblock other pull requests that need rebasing.

**Comment by pwadmore-ea** [comment]
Performing administrative merge override due to failing unit tests for known reasons- see PR approval comments for details.

File: packages/application-to-register-webapp/src/views/layout.html [hunk 27]
```diff
@@ -80,10 +80,9 @@ <h1 class="govuk-heading-xl">Default page template</h1>
 
 {% block bodyEnd %}
   {# Run JavaScript at end of the <body>, to avoid blocking the initial render. #}
-    <script src="{{ assetPath }}/jquery.js"></script>
-    <script src="{{ assetPath }}/all.js"></script>
-    <script src="{{ assetPath }}/moj/all.js"></script>
-    <script>window.GOVUKFrontend.initAll()</script>
-    <script>window.MOJFrontend.initAll()</script>
-    <script src="{{ assetPath }}/js/core.js"></script>
```
Comment (teddmason) [inline packages/application-to-register-webapp/src/views/layout.html:88]: @reneli5 @pwadmore-ea Hi Rene, just found an issue rebasing with this work, do you know why this line for core.js inclusion was removed in your pull request?
